### ISSUE TITLE
lgb.Dataset with free_raw_data = FALSE still raises an error

### ISSUE THREAD

**Issue by mocista** [microsoft/LightGBM#6008]
## Description

Defining a lgb.Dataset with free_raw_data = FALSE. Then slicing it in two parts and using them  as train/validation sets in lightgbm (R). The call "lgb.train(..." fails with an error "   **please set ‘free_raw_data = FALSE’ when you construct lgb.Dataset**". Don't understand why. Can anyone help please?

## Reproducible example

library(lightgbm)

boston = MASS::Boston
str(boston)
dim(boston)

set.seed(12)
boston_lgb_dataset = lgb.Dataset(scale(boston[, -14]), label = boston[,  14] ,free_raw_data = FALSE)

dtrain = lightgbm::slice(boston_lgb_dataset, c(1:350))
dtest = lightgbm::slice(boston_lgb_dataset, c(351:506))

params = list(
  objective = "regression"
  , metric = "l2"
  , min_data = 1L
  , learning_rate = .3
)
 
model = lgb.train( 
  params = params
  , data = dtrain
  , nrounds = 20L
   , valids = list( test  = dtest) 
 )

## Environment info
R version 4.2.0

LightGBM version: 3.3.5

**Comment by walkerjameschris** [comment]
I [opened a PR for this as a possible fix](https://github.com/microsoft/LightGBM/pull/6844).

**Comment by jameslamb** [comment]
Commenting for anyone subscribed to this issue... I think the root cause of this might be that `lgb.Dataset` objects are `{R6}` classes, and `identical()` used to compare 2 such objects is not a good choice for checking "do these Datasets have the same bin mappers?"

https://github.com/microsoft/LightGBM/pull/6844#discussion_r1984482550
